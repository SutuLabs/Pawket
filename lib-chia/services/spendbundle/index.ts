export * from "./defs";
export * from "./validator";
export * from "./signer";
