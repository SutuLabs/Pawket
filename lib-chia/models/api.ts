import { CoinSpend } from "../services/spendbundle";

export interface ApiResponse {
  error?: string;
  success?: boolean;
}

export interface GetParentPuzzleRequest {
  parentCoinId: string;
}

export interface GetParentPuzzleResponse {
  amount: number;
  parentCoinId: string;
  parentParentCoinId: string;
  puzzleReveal: string;
}

export interface GetCoinSolutionResponse {
  coinSpends: CoinSpend[];
}

export interface GetExchangeRateResponse {
  source: string;
  from: string;
  to: string;
  price: number;
  time: string;
}

export interface GetBlockResponse {
  blocks: BlockTransactionGeneratorRetrieval[];
  refBlocks: BlockTransactionGeneratorRetrieval[];
}

export interface BlockTransactionGeneratorRetrieval {
  index: number;
  generator: string;
  generator_ref_list: number[];
}
