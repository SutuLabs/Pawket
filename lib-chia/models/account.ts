import { AccountKey } from "../services/crypto/account";
import { AddressType } from "../services/crypto/puzzle";
import { DidDetail, NftDetail, TokenPuzzleDetail, TokenPuzzleObserver } from "../services/crypto/receive";
import { DonwloadedNftCollection } from "./nft";
import { CoinRecord } from "./wallet";

export type AccountType = "Serial" | "Password" | "Legacy" | "Address" | "PublicKey";

export interface AccountTokenAddress {
  address: string;
  type?: AddressType;
  coins: CoinRecord[];
}

export interface AccountToken {
  amount: bigint;
  addresses: AccountTokenAddress[];
}
export interface AccountTokens {
  [symbol: string]: AccountToken;
}

export interface PersistentAccount {
  key: AccountKey;
  name: string;
  profilePic?: string;
  type: AccountType;
  serial?: number;
  puzzleHash?: string;
  addressRetrievalCount: number;
  cats?: CustomCat[];
  allCats: PersistentCustomCat[];
}

export interface AccountEntity extends PersistentAccount {
  firstAddress?: string;
  activities?: CoinRecord[];
  tokens: AccountTokens;
  nfts?: NftDetail[];
  extraInfo: DonwloadedNftCollection;
  dids?: DidDetail[];
  addressPuzzles: TokenPuzzleDetail[];
  observePuzzles?: TokenPuzzleObserver[];
  addressGenerated: number;
}

export interface CustomCat {
  name: string;
  id: string;
  img?: string;
}

export interface PersistentCustomCat extends CustomCat {
  network: string;
}

export interface TokenInfo {
  [symbol: string]: OneTokenInfo;
}

export interface OneTokenInfo {
  id?: string;
  symbol: string;
  decimal: number;
  unit: string;
  img?: string;
}
