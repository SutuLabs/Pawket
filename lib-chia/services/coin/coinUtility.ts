import { Hex0x, prefix0x } from "./condition";
import { Bytes, bigint_to_bytes } from "clvm";
import { GetPuzzleApiCallback } from "../transfer/transfer";
import { CoinItem, CoinRecord } from "../../models/wallet";
import { CoinSpend, OriginCoin } from "../spendbundle";
import { AccountEntity } from "../../models/account";

export interface NetworkContext {
  prefix: string;
  symbol: string;
  chainId: string;
  api: GetPuzzleApiCallback;
}

export interface NetworkContextWithOptionalApi {
  prefix: string;
  symbol: string;
  chainId: string;
  api?: GetPuzzleApiCallback;
}

export interface CompatibleCoin {
  amount: bigint | number;
  parent_coin_info: Hex0x | string;
  puzzle_hash: Hex0x | string;
}

export interface LockedCoin {
  coinName: string;
  coin: OriginCoin;
  transactionTime: number;
  network: string;
  symbol?: string;
  accountFinger: number;
}

export interface PendingTransaction {
  coin: LockedCoin[];
  time: number;
  network: string;
  amount: { [key: string]: bigint };
}

export async function lockCoins(
  account: AccountEntity,
  coinSpends: CoinSpend[],
  transactionTime: number,
  network: string
): Promise<void> {
  const lc: LockedCoin[] = getLockedCoinsFromLocalStorage();
  const accountFinger = account.key.fingerprint;
  for (const cs of coinSpends) {
    const coinName = getCoinName(cs.coin);
    const idx = account.activities?.findIndex((act) => act.coin?.puzzleHash == cs.coin.puzzle_hash);
    if (idx != undefined && idx > -1)
      lc.push({
        coinName: coinName,
        coin: cs.coin,
        transactionTime: transactionTime,
        network: network,
        symbol: account.activities && account.activities[idx].symbol,
        accountFinger: accountFinger,
      });
  }
  localStorage.setItem("LOCKED_COINS", JSON.stringify(lc));
}

export function unlockCoins(coins: OriginCoin[]): OriginCoin[] {
  const lc: LockedCoin[] = getLockedCoinsFromLocalStorage();
  const uc: OriginCoin[] = [];
  for (const coin of coins) {
    if (!coin) continue;
    const cname = getCoinName(coin);
    const idx = lc.findIndex((c) => c.coinName == cname);
    if (idx > -1) {
      uc.push(coin);
      lc.splice(idx, 1);
    }
  }
  localStorage.setItem("LOCKED_COINS", JSON.stringify(lc));
  return uc;
}

export function getCompletedTransactions(coins: OriginCoin[], activities: CoinRecord[]): number[] {
  const txs: number[] = [];
  for (const sAct of activities) {
    if (sAct.spent && sAct.coin) {
      const idx = coins.findIndex(
        (c) => c.puzzle_hash == sAct.coin?.puzzleHash && c.parent_coin_info == sAct.coin.parentCoinInfo
      );
      if (idx > -1) txs.push(sAct.spentBlockIndex);
    }
  }
  return txs;
}

export function coinFilter(account: AccountEntity, coins: OriginCoin[], network: string): OriginCoin[] {
  let lc: LockedCoin[] = getLockedCoinsFromLocalStorage();
  lc = lc.filter((l) => l.network == network && l.accountFinger == account.key.fingerprint);
  return coins.filter((coin) => {
    const name = getCoinName(coin);
    return lc.findIndex((c) => c.coinName == name) == -1;
  });
}

export function getLockedCoinsFromLocalStorage(): LockedCoin[] {
  const lcStr = localStorage.getItem("LOCKED_COINS");
  let lc: LockedCoin[];
  try {
    lc = lcStr ? JSON.parse(lcStr) : [];
  } catch (error) {
    lc = [];
  }
  return lc;
}

export function getCoinName0x(coin: CompatibleCoin): Hex0x {
  return prefix0x(getCoinNameHex(coin).hex());
}

export function getCoinName(coin: CompatibleCoin): string;
export function getCoinName(coin: undefined): undefined;
export function getCoinName(coin: CompatibleCoin | undefined): string | undefined;
export function getCoinName(coin: CompatibleCoin | undefined): string | undefined {
  if (!coin) return coin;
  return getCoinNameHex(coin).hex();
}

export function getCoinNameHex(coin: CompatibleCoin): Bytes {
  const a = bigint_to_bytes(BigInt(coin.amount), { signed: true });
  const pci = Bytes.from(coin.parent_coin_info, "hex");
  const ph = Bytes.from(coin.puzzle_hash, "hex");
  const cont = pci.concat(ph).concat(a);
  const coinname = Bytes.SHA256(cont);
  return coinname;
}

export function convertToOriginCoin(
  coin: CoinItem | { amount: number; parent_coin_info: Hex0x | string; puzzle_hash: Hex0x | string }
): OriginCoin {
  return "parentCoinInfo" in coin
    ? {
        amount: BigInt(coin.amount),
        parent_coin_info: coin.parentCoinInfo,
        puzzle_hash: coin.puzzleHash,
      }
    : {
        amount: BigInt(coin.amount),
        parent_coin_info: prefix0x(coin.parent_coin_info),
        puzzle_hash: prefix0x(coin.puzzle_hash),
      };
}
