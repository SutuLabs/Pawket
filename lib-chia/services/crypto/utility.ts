import { PrivateKey, G1Element, ModuleInstance } from "@chiamine/bls-signatures";
import { prefix0x, unprefix0x } from "../coin/condition";
import { Bytes } from "clvm";
import crypto from "./isoCrypto";
import { Instance } from "../util/instance";

type deriveCallback = (path: number[]) => PrivateKey;
type derivePkCallback = (path: number[]) => G1Element;

class Utility {
  toHexString(byteArray: Uint8Array) {
    return Array.from(byteArray, function (byte) {
      return ("0" + (byte & 0xff).toString(16)).slice(-2);
    }).join("");
  }

  fromHexString(hexString: string | undefined): Uint8Array {
    hexString = unprefix0x(hexString);
    if (!hexString) return new Uint8Array();
    return new Uint8Array(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (<any>hexString).match(/.{1,2}/g).map((byte: any) => parseInt(byte, 16))
    );
  }

  derivePath(BLS: ModuleInstance, sk: PrivateKey, path: number[], hardened: boolean): PrivateKey {
    for (let i = 0; i < path.length; i++) {
      const p = path[i];
      sk = hardened ? BLS.AugSchemeMPL.derive_child_sk(sk, p) : BLS.AugSchemeMPL.derive_child_sk_unhardened(sk, p);
    }
    return sk;
  }

  async derive(privateKey: Uint8Array, hardened = true): Promise<deriveCallback> {
    const BLS = Instance.BLS;
    if (!BLS) throw new Error("BLS not initialized");
    const sk = BLS.PrivateKey.from_bytes(privateKey, false);
    return (path: number[]) => this.derivePath(BLS, sk, path, hardened);
  }

  derivePkPath(BLS: ModuleInstance, pk: G1Element, path: number[]): G1Element {
    for (let i = 0; i < path.length; i++) {
      const p = path[i];
      pk = BLS.AugSchemeMPL.derive_child_pk_unhardened(pk, p);
    }
    return pk;
  }

  async derivePk(publicKey: Uint8Array): Promise<derivePkCallback> {
    const BLS = Instance.BLS;
    if (!BLS) throw new Error("BLS not initialized");
    const pk = BLS.G1Element.from_bytes(publicKey);
    return (path: number[]) => this.derivePkPath(BLS, pk, path);
  }

  async getPrivateKey(privateKey: Uint8Array): Promise<PrivateKey> {
    const BLS = Instance.BLS;
    if (!BLS) throw new Error("BLS not initialized");
    const sk = BLS.PrivateKey.from_bytes(privateKey, false);
    return sk;
  }

  public async purehash(data: string | ArrayBuffer): Promise<Uint8Array> {
    if (typeof data === "string") {
      const enc = new TextEncoder();
      return new Uint8Array(await crypto.subtle.digest("SHA-256", enc.encode(data)));
    } else {
      return new Uint8Array(await crypto.subtle.digest("SHA-256", data));
    }
  }

  public async hash(data: string): Promise<string> {
    return this.toHexString(await this.purehash(data));
  }

  public async getFileHash(f: File): Promise<string> {
    const buffer = await f.arrayBuffer();
    const hash = await crypto.subtle.digest("SHA-256", buffer);
    return this.toHexString(new Uint8Array(hash));
  }

  public async strongHash(data: string, salt: string, iteration = 1024): Promise<string> {
    // const start = performance.now();
    const enc = new TextEncoder();
    let d = Uint8Array.from([...enc.encode(data), ...enc.encode(salt)]).buffer;
    for (let i = 0; i < iteration; i++) {
      d = await crypto.subtle.digest("SHA-256", d);
    }
    // const end = performance.now();
    // console.warn(`elapsed ${end - start}ms for ${iteration} iterations.`);
    return this.toHexString(new Uint8Array(d));
  }

  public getRandom(length: number): Uint8Array {
    const seed = new Uint8Array(length);
    crypto.getRandomValues(seed);
    return seed;
  }
}

export default new Utility();

export function bytesToHex0x(bytes: Bytes): string {
  return prefix0x(bytes.hex());
}
