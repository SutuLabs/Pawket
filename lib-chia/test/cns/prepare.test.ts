import { getTestAccount } from "../utility";
import { SymbolCoins } from "../../services/transfer/transfer";
import { getBootstrapSpendBundle } from "../../services/coin/nft";
import { GetParentPuzzleResponse } from "../../models/api";
import { Instance } from "../../services/util/instance";
import { getAccountAddressDetails } from "../../services/util/account";

import { knownCoins } from "./cns.test.data";
import { NetworkContext } from "../../services/coin/coinUtility";

import { assertSpendbundle } from "../../services/spendbundle/validator";
import { signSpendBundle } from "../../services/spendbundle";

const net: NetworkContext = {
  prefix: "xch",
  symbol: "XCH",
  chainId: "ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb",
  api: localPuzzleApiCall,
};
function xchPrefix() {
  return "xch";
}
function xchSymbol() {
  return "XCH";
}
function _chainId() {
  return "ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb";
}
function tokenInfo() {
  return {};
}

beforeAll(async () => {
  await Instance.init();
});

test("Prepare CNS bootstrap coins", async () => {
  const fee = 0n;
  const availcoins: SymbolCoins = {
    [xchSymbol()]: [
      {
        amount: 23n,
        parent_coin_info: "0xc4badc175d119df8006fd8e96ad84c475e743275e70d6c16f43cf83fb75df021",
        puzzle_hash: "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
      },
      {
        amount: 4998999984n,
        parent_coin_info: "0xf3b7d6d4bdd80b99c539f7ca900288f5dc2ac8fb23559656e981761e90b2fe71",
        puzzle_hash: "0x0eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155",
      },
    ],
  };
  const count = 10;

  const target_hex = "0x0eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155";
  const change_hex = "0x0eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155";

  const account = getTestAccount("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");

  const tokenPuzzles = await getAccountAddressDetails(account, [], tokenInfo(), xchPrefix(), xchSymbol(), undefined, "cat_v2");
  const sk = "00186eae4cd4a3ec609ca1a8c1cda8467e3cb7cbbbf91a523d12d31129d5f8d7";

  const ubundle = await getBootstrapSpendBundle(target_hex, change_hex, fee, availcoins, tokenPuzzles, count, net, sk);
  const bundle = await signSpendBundle(ubundle, tokenPuzzles, net.chainId);
  await assertSpendbundle(bundle, net.chainId);
  expect(bundle).toMatchSnapshot("spendbundle");
});

async function localPuzzleApiCall(parentCoinId: string): Promise<GetParentPuzzleResponse | undefined> {
  const resp = knownCoins.find((_) => _.parentCoinId == parentCoinId);
  return resp;
}
