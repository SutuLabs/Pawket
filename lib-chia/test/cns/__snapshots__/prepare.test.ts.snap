// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prepare CNS bootstrap coins: spendbundle 1`] = `
Object {
  "aggregated_signature": "0x85b2fcf01bb343eb0232901698eeb8656c749d9bcb2f7a2ef41b7943644269472a62ac962cd643378f8394bab1e745de055e8bdc44fabbd561943014016aaf7542b9c474c151d853e1d84453c9c023b18b5671fdb4aef58ce796988a011a1d48",
  "coin_spends": Array [
    Object {
      "coin": Object {
        "amount": 23n,
        "parent_coin_info": "0xc4badc175d119df8006fd8e96ad84c475e743275e70d6c16f43cf83fb75df021",
        "puzzle_hash": "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b092c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1dff018080",
      "solution": "0xff80ffff01ffff33ffa0d4b6b47b409215438ea21aaee024978e3f77be69703c8d0c0b56d2aa1bf48514ff0180ffff33ffa07df1f15126986137ec536c9622bc0e179e003eaacb80f414c8e9bdc6c78ec93eff0180ffff33ffa018790cc0307580f97ab574790b5d6df468d0731ff22b1cb84633918ef8bc6d08ff0180ffff33ffa076e217ca2e841e9dbf9ae3a781d5fcb785b6ba7e5cff3539a1b691b8c5f39414ff0180ffff33ffa04d75e2db26f791397f5852331917b0f80df48086787aeddbed609b1f1ecfbae3ff0180ffff33ffa02a9a38c475d39d7d08d215621c215004cd3803167a914e2dde3ca54c92f565b9ff0180ffff33ffa025345ffa379e2fddfdd8ec5a1f73d4673db3c13b9cfb455741293a5c7a0d00ffff0180ffff33ffa0523f6dd8d5e93ccc91f35762012b06704ec58bf5ded360169e0eda9a2b4ab5a5ff0180ffff33ffa0461f88c07af2d21f41ef7be9e159d492843f197e11c780d1c9a9bb477dd31c71ff0180ffff33ffa0fd51a8082b3b1388d325464a090c4a269866dbab4ce716535b2bbf9006e392a5ff0180ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff0d8080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0xd4b6b47b409215438ea21aaee024978e3f77be69703c8d0c0b56d2aa1bf48514",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b082f99159ffdf03708b449d52e5eb7dafde7f57a2d6635fbccdb1932b9fd1d73163feeaa1ddad6b93b4d933a6bbee3a5aff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x7df1f15126986137ec536c9622bc0e179e003eaacb80f414c8e9bdc6c78ec93e",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b095c42d5bed266bbaa0916e5477e3278c4fab6341583e91add000bfd8d374cf8aa313d89aea6782050c05963e3a1c3a44ff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x18790cc0307580f97ab574790b5d6df468d0731ff22b1cb84633918ef8bc6d08",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0a73f8eaaebde20e4758c877f3ab659079e0e4e835ba99023b9ad018fba16de8139cb9209467656e79060c04058369060ff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x76e217ca2e841e9dbf9ae3a781d5fcb785b6ba7e5cff3539a1b691b8c5f39414",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0ab197e05c1c475cce2e9232f2bdf69e82d84c492607da30e6847e86dec898635368c4e9e658e9096e117d21206e922baff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x4d75e2db26f791397f5852331917b0f80df48086787aeddbed609b1f1ecfbae3",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b09761c3e95673f8593a65418ecb4457ffa3ff207099d77cd6e30102eed8d451ca4cacdf1baccb21750f8b98e9adb3633dff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x2a9a38c475d39d7d08d215621c215004cd3803167a914e2dde3ca54c92f565b9",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b081da07b52d4e08df48dea15bd9189bef12843149338071c9e6edae6d8c0c811e67c02932f98a2d46ce083d84812840e2ff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x25345ffa379e2fddfdd8ec5a1f73d4673db3c13b9cfb455741293a5c7a0d00ff",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b08d9a05678bbf80631497f72cd5b92e8a63a19ddcba76447bfaff37ac3c746825342fbd4ad1b0af399d43d009d07475aeff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x523f6dd8d5e93ccc91f35762012b06704ec58bf5ded360169e0eda9a2b4ab5a5",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b081675f0b9e91116b013bb35ee53ebb349be651684ebb4d6c9c0ca18ff8898eb9530e6f840079db23f8fe26cc453341d8ff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0x461f88c07af2d21f41ef7be9e159d492843f197e11c780d1c9a9bb477dd31c71",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0b667d74db6caa1e6c1154435800bc7baeed1d06a870d7ee427d3b5660e4722dfbe13c2e2c0cb1fd837240b086405a458ff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x18494efb9385b2604bb168f070540d42e8133f3b7204ab74896d8c21a6c73cc0",
        "puzzle_hash": "0xfd51a8082b3b1388d325464a090c4a269866dbab4ce716535b2bbf9006e392a5",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b0a15aae698d29e55475704328beec7a77ec48ecf93bc5674ef320b274bfa24e0b652ec704b8652d839d4efc2931d91afcff018080",
      "solution": "0xff80ffff01ffff33ffa00eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155ff018080ff8080",
    },
  ],
}
`;
